runtime/limits_integ_test.go

```diff
@@ -0,0 +1,174 @@
+	require.NoError(err, "unable to create client to containerd service at %s, is containerd running?", containerdSockPath)
+	defer client.Close()
+
+	image, err := alpineImage(ctx, client, naiveSnapshotterName)
+	require.NoError(err, "failed to get alpine image")
+
+	sh := containerd.WithNewSpec(oci.WithProcessArgs("sh", "/var/firecracker-containerd-test/scripts/diskhog.sh"),
+		oci.WithDefaultPathEnv,
+		oci.WithMounts([]specs.Mount{
+			{
+				Type:        "bind",
+				Options:     []string{"bind"},
+				Destination: "/var/firecracker-containerd-test/scripts",
+				Source:      "/var/firecracker-containerd-test/scripts",
+			},
+		}))
+
+	container, err := client.NewContainer(ctx,
+		"container",
+		containerd.WithSnapshotter(naiveSnapshotterName),
+		containerd.WithNewSnapshot("snapshot", image),
+		sh,
+	)
+	defer func() {
+		err = container.Delete(ctx, containerd.WithSnapshotCleanup)
+		require.NoError(err, "failed to delete a container")
+	}()
+
+	result, err := runTask(ctx, container)
+	require.NoError(err, "failed to create a container")
+
+	assert.Equal(uint32(1), result.exitCode, "exitCode")
+
+	assert.Equal(`Filesystem                Size      Used Available Use% Mounted on
+/dev/vdb                975.9M      8.4M    900.3M   1% /
+`, result.stdout, "stdout")
+
+	// TODO: Doesn't it have only 900MB? The written records are higher than expected.
+	assert.Equal(`952+0 records in
+951+0 records out
+`, result.stderr, "stderr")
+}
+
+func TestMemoryLimit_Isolated(t *testing.T) {
+	internal.RequiresIsolation(t)
+
+	assert := assert.New(t)
+	require := require.New(t)
+
+	ctx := namespaces.WithNamespace(context.Background(), "default")
+
+	client, err := containerd.New(containerdSockPath, containerd.WithDefaultRuntime(firecrackerRuntime))
+	require.NoError(err, "unable to create client to containerd service at %s, is containerd running?", containerdSockPath)
+	defer client.Close()
+
+	image, err := alpineImage(ctx, client, naiveSnapshotterName)
+	require.NoError(err, "failed to get alpine image")
+
+	sh := containerd.WithNewSpec(oci.WithProcessArgs("sh", "/var/firecracker-containerd-test/scripts/memoryhog.sh"),
+		oci.WithDefaultPathEnv,
+		oci.WithMounts([]specs.Mount{
+			{
+				Type:        "bind",
+				Options:     []string{"bind"},
+				Destination: "/var/firecracker-containerd-test/scripts",
+				Source:      "/var/firecracker-containerd-test/scripts",
+			},
+		}),
+	)
+
+	container, err := client.NewContainer(ctx,
+		"container",
+		containerd.WithSnapshotter(naiveSnapshotterName),
+		containerd.WithNewSnapshot("snapshot", image),
+		sh,
+	)
+	defer func() {
+		err = container.Delete(ctx, containerd.WithSnapshotCleanup)
+		require.NoError(err, "failed to delete a container")
+	}()
+
+	result, err := runTask(ctx, container)
+	require.NoError(err, "failed to create a container")
+
+	assert.Equal(uint32(1), result.exitCode)
+	assert.Equal("", result.stdout, "stdout")
+	assert.Equal("sh: out of memory\n", result.stderr, "stderr")
```

Comment:
I don't see a memory limit specified anywhere.  Where are you getting a limit from?

Reply:
This limit is coming from Firecracker's default memory limit. I would have an explicit Create VM Request.

Reply:
Putting it in the CreateVM request would help a lot.  I think another useful thing would be to assert on the amount of memory the `firecracker` process is using (as a proxy for the amount of memory used by the VM).  There would need to be some amount of buffer for the overhead of Firecracker itself, but that way we'd be testing the VMM and our configuration of it rather than the behavior of the Linux kernel inside the VM.

Reply:
I agree with Sam but would go further and say that all firecracker-containerd code does is translate CreateVM API parameters into Firecracker API calls, so I feel like an e2e test of just *our* code would make a CreateVM call with a given memory limit and then open a client directly with the VMM and verify it reports it was actually configured with the expected memory limit. Verifying the memory limit was passed through as expected is what we actually care about testing, not that the Firecracker implementation is also honoring that memory limit (presumably they already have tests for that in their own repo).

Now, if it turns out that for some reason it's actually harder to ask the VMM directly for what memory limit it was configured with compared to asserting on the VMM process's memory usage, then I'd be fine with asserting on the VMM process's memory usage. I just highly suspect that the issue Sam brought up with needing to deal with the VMM's own memory usage introduces enough complication to that approach that it's probably easier to write a non-flaky test that just talks to the VMM API endpoint to make assertions on the outcome of a CreateVM call.

Reply:
Yeah. I agree that this test is doing what Firecracker supposed to do. Firecracker has some tests like https://github.com/firecracker-microvm/firecracker/commit/36583c62714c5a5d5637ea2df0cce6624d0d7b46 to confirm the behavior.

Checking only the data transform layer would not cover handling of stdout, stderr and/or exit code, but those are technically covered by the disk usage test I have in this PR.

Reply:
I have removed that from this PR. Still not so sure what would be the best way to confirm the behavior.

We may also set the limit through https://github.com/opencontainers/runtime-spec/blob/master/config-linux.md#memory.